internal/pkg/term/progress/render.go

```diff
@@ -0,0 +1,44 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package progress
+
+import (
+	"bytes"
+	"io"
+)
+
+// Renderer is the interface to print a component to a writer.
+// It returns the number of lines printed and the error if any.
+type Renderer interface {
+	Render(out io.Writer) (numLines int, err error)
+}
+
+// allOrNothingRenderer renders all partial renders or none of them.
+type allOrNothingRenderer struct {
+	err      error
+	numLines int
+	buf      bytes.Buffer
+}
+
+func (r *allOrNothingRenderer) Partial(renderer Renderer) {
```

Comment:
They can be private! The only advantage of having `Render` instead of `render` is that `allOrNothingRenderer` gets to satisfy the `Renderer` interface that way.
So you can do crazy things like mix an `allOrNothingRenderer` with a `singleLineComponent`